gapic-generator/test/google/gapic/annotations/method_test.rb

```diff
@@ -0,0 +1,242 @@
+
+    refute_nil method.http
+    assert_kind_of Google::Api::HttpRule, method.http
+    assert_equal "/v1/simple_garbage:get", method.http.post
+    assert_equal "*", method.http.body
+
+    assert_empty method.signatures
+
+    assert_nil method.operation_info
+  end
+
+  def test_garbage_GetSpecificGarbage
+    garbage = api :garbage
+    service = garbage.services.find { |s| s.name == "GarbageService" }
+    refute_nil service
+    method = service.methods.find { |s| s.name == "GetSpecificGarbage" }
+    refute_nil method
+
+    assert_kind_of Google::Protobuf::MethodOptions, method.options
+    assert_kind_of Google::Api::HttpRule, method.options[:http]
+    assert_equal "/v1/specific_garbage:get", method.options[:http][:post]
+    assert_equal "*", method.options[:http][:body]
+
+    expected_signatures = [
+      "name,int32,bool",
+    ]
+    assert_equal expected_signatures, method.options[:method_signature]
+
+    assert_nil method.options[:operation_info]
+
+    assert_nil method.operation_info
+
+    refute_nil method.http
+    assert_kind_of Google::Api::HttpRule, method.http
+    assert_equal "/v1/specific_garbage:get", method.http.post
+    assert_equal "*", method.http.body
+
+    assert_equal expected_signatures, method.signatures
+
+    assert_nil method.operation_info
+  end
+
+  def test_garbage_GetNestedGarbage
+    garbage = api :garbage
+    service = garbage.services.find { |s| s.name == "GarbageService" }
+    refute_nil service
+    method = service.methods.find { |s| s.name == "GetNestedGarbage" }
+    refute_nil method
+
+    assert_kind_of Google::Protobuf::MethodOptions, method.options
+    assert_kind_of Google::Api::HttpRule, method.options[:http]
+    assert_equal "/v1/nested_garbage:get", method.options[:http][:post]
+    assert_equal "*", method.options[:http][:body]
+
+    expected_signatures = [
+      "name,int32,int64",
+      "name,float,double",
+      "name,bool,enum"
+    ]
+    assert_equal expected_signatures, method.options[:method_signature]
+
+    assert_nil method.options[:operation_info]
+
+    assert_nil method.operation_info
+
+    refute_nil method.http
+    assert_kind_of Google::Api::HttpRule, method.http
+    assert_equal "/v1/nested_garbage:get", method.http.post
+    assert_equal "*", method.http.body
+
+    assert_equal expected_signatures, method.signatures
+
+    assert_nil method.operation_info
+  end
+
+  def test_garbage_GetRepeatedGarbage
+    garbage = api :garbage
+    service = garbage.services.find { |s| s.name == "GarbageService" }
+    refute_nil service
+    method = service.methods.find { |s| s.name == "GetRepeatedGarbage" }
+    refute_nil method
+
+    assert_kind_of Google::Protobuf::MethodOptions, method.options
+    assert_kind_of Google::Api::HttpRule, method.options[:http]
+    assert_equal "/v1/repeated_garbage:get", method.options[:http][:post]
+    assert_equal "*", method.options[:http][:body]
+
+    expected_signatures = [
+      "name,bool,enum",
```

Comment:
Might not be a bad idea to add some nested signatures in one of these tests. i.e. `name,bool,foo.bar.baz`

Reply:
I don't understand how that would work. How do you define nested parameters to a Ruby method?

Reply:
Meaning a nested field in a message.

example:

```proto
rpc GetThing(ThingRequest) returns (ThingResponse) {
    option (google.api.method_signature) = "top_level_field,submessage.nested_field";
}

message ThingRequest {
  string top_level_field = 1;
  Submessage submessage = 2;
}

message Submessage {
  string nested_field = 1;
}
```

this should yield a method whose signature can be called like:

```rb
client.get_thing(top_level_field: "top", nested_field: "nested")

# or this depending on how we want to handle some disambiguation cases

client.get_thing(top_level_field: "top", submessage_nested_field: "nested")

Reply:
In that case the signature would be `top_level_field,submessage`, wouldn't it? The way you would call it is:

```ruby
client.get_thing(top_level_field: "top", submessage: { nested_field: "nested" })
```

We've talked about not even using these signatures outside of possibly documentation code examples. I don't know how we would support dotted field names in ruby method arguments, as Ruby does not have method overloading like other languages, but there is no reason dotted field names can't exist in the annotations.

Reply:
>In that case the signature would be top_level_field,submessage, wouldn't it? 
Nope the new annotations will allow for `top_level_field,submessage.nested_field`

Can you give an example of how dotted names couldn't work?

Reply:
Can we table this for now and address it in a separate issue?

Reply:
Let's table how we support nested fields but can you add the nested field to the annotation since the it is a valid input.

Reply:
You want me to add an annotation for a nested field even though we don't have any nested fields to reference? So a method signature of `name,foo.bar.baz`?

Reply:
Added.

Reply:
Ah sorry I was unclear. To be more specific. Heres the changes:

```diff
rpc GetSpecificGarbage(SpecificGarbage) returns (SpecificGarbage) {
  ...
  option (google.api.method_signature) = "name,int32,bool";
+ option (google.api.method_signature) = "name,nested.int64";
}

message SpecificGarbage {
  string name = 1 [(google.api.field_behavior) = REQUIRED];
  ...
  GarbageEnum enum = 11;
+ NestedGarbage nested = 12;

  message NestedGarbage {
    ...
    int64 int64 = 3;
    ...
  }
}
```

Reply:
Added.